source/retryable-reads/retryable-reads.rst

```diff
@@ -331,6 +331,13 @@ other error originating from the server, that error should be raised instead as
 the caller can infer that an attempt was made and the second error is likely
 more relevant (with respect to the current topology state).
 
+If a driver associates server information (e.g. the server address or
+description) with an error, the driver MUST ensure the server information
+reported with that error corresponds to the server that was selected when the
+error occurred. Specifically, if a retry attempt fails, the server reported
+with the error MUST correspond to the server that was selected for the retry
+and MUST NOT simply be carried over from the original error.
```

Comment:
This last sentence seems at odds with the following language in the description of DRIVERS-2170:
> I think they could implement this ticket just by ensuring that the attached server is always the originating server of the error being reported. For example, a driver that always attaches the first attempt's server to such an exception should be changed to conditionally attach the first or second accordingly.

Can we clarify here that the server reported should be the one that corresponds to the error that gets returned to the user?

Reply:
I don't see a conflict between this text and the JIRA description. The JIRA issue was requesting that any server info attached to a retry attempt error indicate the server used for that retry attempt. "Server that was selected for the retry" suggests the same thing.

What if we replace this entire paragraph with:

> If a driver associates server information (e.g. the server address or description) with an error, the driver MUST ensure that the reported server information corresponds to the server that originated the error.

This does away with any reference to "selected" and instead focuses on what was used for the operation or retry attempt.

----

To @kkloberdanz's earlier comment (https://github.com/mongodb/specifications/pull/1480#issuecomment-1856576862), I don't think there's any syntax in the unified test format that would be useful for testing this (see: [expectedError](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#expectederror)).

A possible prose test would entail comparing server information reported in a driver exception to the servers inferred from command monitoring events for the original and second attempt. But any meaningful test would also require setting up fail points on two _different_ servers, since consecutive errors on the same host (e.g. failing `insert` twice on a primary) would behavior the same irrespective of this spec language.

We intentionally don't have any prose tests that deal with failovers (Atlas failover testing is its own mess), and I don't think that's worth the trouble for this spec change.

[DRIVERS-2170](https://jira.mongodb.org/browse/DRIVERS-2170) was prompted by my own personal pedantry (Ruby historically did The Right Thing™) and I sincerely regret the sorrow, wailing, and gnashing of teeth it has caused among DBX.

Reply:
I like the simplified paragraph. I'll make that change.